ui/src/saros/ui/browser/SWTBrowser.java

```diff
@@ -0,0 +1,199 @@
+package saros.ui.browser;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import org.eclipse.swt.browser.Browser;
+import org.eclipse.swt.browser.BrowserFunction;
+import org.eclipse.swt.browser.ProgressEvent;
+import org.eclipse.swt.browser.ProgressListener;
+import org.eclipse.swt.events.DisposeEvent;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Display;
+
+public class SWTBrowser implements IBrowser {
+
+  private enum ReadyState {
+    LOADING,
+    COMPLETE
+  }
+
+  private final Browser browser;
+  private ReadyState state = ReadyState.LOADING;
+  private List<Runnable> runOnDisposalList =
+      Collections.synchronizedList(new ArrayList<Runnable>());
+
+  private final ProgressListener readyStateListener =
+      new ProgressListener() {
+        @Override
+        public void completed(ProgressEvent event) {
+          state = ReadyState.COMPLETE;
+        }
+
+        @Override
+        public void changed(ProgressEvent event) {
+          state = ReadyState.LOADING;
+        }
+      };
+
+  public SWTBrowser(Composite parent, int style) {
+    browser = new Browser(parent, style);
+
+    browser.addDisposeListener(
+        (DisposeEvent e) -> {
+          synchronized (runOnDisposalList) {
+            for (Runnable runnable : runOnDisposalList) {
+              runnable.run();
```

Comment:
When one runnable raises a runtime exception nothing will be executed afterwards. Is this intended ?

Reply:
Done